### PR TITLE
[csrng/rtl] internal state read timing improvements

### PR DESCRIPTION
**Pull request by mwbranstad**
Re-structed the read flow for the internal state array.
Added parameter to prim_fifo_sync to remove output mux.
Added simple flops to break up configuration paths.

Signed-off-by: Mark Branstad <mark.branstad@wdc.com>

### CODE REVIEW

File: hw/ip/aon_timer/data/aon_timer.hjson
Comment (vogelpi) [inline hw/ip/aon_timer/data/aon_timer.hjson:221]: This change seems to be unrelated to csrng. It should go into a separate PR.

File: hw/ip/csrng/rtl/csrng_cmd_stage.sv
Comment (vogelpi) [inline hw/ip/csrng/rtl/csrng_cmd_stage.sv:325]: You still have one FIFO here with the output mux. Could be a typo or on purpose, I can't judge. If it's on purpose, it would be good to a comment to motivate the choice.

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (vogelpi) [inline hw/ip/csrng/rtl/csrng_state_db.sv:117]: By removing two whitespaces on this line (between `assign` and `internal...`) you can avoid all the changes on the next 15 lines and still keep the statement well aligned.

Comment (vogelpi) [reply hw/ip/csrng/rtl/csrng_state_db.sv:117]: Actually, this could be written as
```sv
always_comb begin
  internal_state_pl_d = 0;
  for (int i = 0; i < MaxNApps; i++) begin
    internal_state_pl_d |= internal_states_out[i];
  end
end
```
Don't know if the syntax is correct. But it would allow you to avoid the hardcoded literals that need to be updated if changing MaxNApps.

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (vogelpi) [inline hw/ip/csrng/rtl/csrng_state_db.sv:113]: Not sure if this is a potential issue: previously, the state read out was gated by lc here. Now you just do it on the read mux, but the state still ends up inside the `internal_state_pl_q` flops. Adding the lc signal as a write enable for the flops wouldn't hurt I think.

**Review by vogelpi — COMMENTED** [review]
Thanks @mwbranstad for looking into this! The changes look mostly good to me, I have a few comments.

The change in the aon timer even if its simple should be factored out into a separate PR as it is not related to csrng.

For the record, this PR is related to #4451.

**Comment by mwbranstad** [comment]
Reviews can be put on hold until I removed the aon_timer.hjson change, as this will go into another PR.

File: hw/ip/aon_timer/data/aon_timer.hjson
Comment (mwbranstad) [reply hw/ip/aon_timer/data/aon_timer.hjson:221]: Done - #5373

File: hw/ip/csrng/rtl/csrng_cmd_stage.sv
Comment (mwbranstad) [reply hw/ip/csrng/rtl/csrng_cmd_stage.sv:325]: On-purpose - reason is that an output assert will fire since X's propagate out. Since I have heard both sides of this debate as the correct solution for x's versus mux 0's, leaving it this way for now, and will leave a comment.

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (mwbranstad) [reply hw/ip/csrng/rtl/csrng_state_db.sv:117]: excellent tip @vogelpi ! This code works, both compiled and functionally. It is very terse, almost unreadable, but I like it. This also removes the need for MaxNApp parameter in general, so overall the code is cleaned up.

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (mwbranstad) [reply hw/ip/csrng/rtl/csrng_state_db.sv:113]: Nope - lc_state should only gate the reading - this is a debug feature so the lc_state was added for security reasons. Otherwise, this array pipeline is the full functional path and should have no influence from lc_state. Thanks for double checking though.

**Comment by mwbranstad** [comment]
Any other feedback on this?

File: hw/ip/aon_timer/data/aon_timer.hjson
Comment (msfschaffner) [reply hw/ip/aon_timer/data/aon_timer.hjson:221]: Is this still needed after the fix by @weicaiyang ?

File: hw/ip/csrng/data/csrng.hjson
Comment (msfschaffner) [inline hw/ip/csrng/data/csrng.hjson:225]: nit: Setting this bit halt**s**...

File: hw/ip/csrng/rtl/csrng_cmd_stage.sv
Comment (msfschaffner) [reply hw/ip/csrng/rtl/csrng_cmd_stage.sv:325]: +1 for adding a comment here

**Review by msfschaffner — COMMENTED** [review]
LGTM with a few minor comments.

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (vogelpi) [reply hw/ip/csrng/rtl/csrng_state_db.sv]: Cool, glad you like it ;-)

File: hw/ip/csrng/rtl/csrng_state_db.sv
Comment (vogelpi) [reply hw/ip/csrng/rtl/csrng_state_db.sv:113]: Okay, this sounds good to me. Thanks for the clarification.

**Review by vogelpi — COMMENTED** [review]
Thanks for addressing my comments @mwbranstad !

I think the PR needs to be rebased on current master. This will allow you to drop the aon_timer change. Plus there is a comment you wanted to add. Once this is done, I am more than happy to approve and merge this PR.

File: hw/ip/aon_timer/data/aon_timer.hjson
Comment (mwbranstad) [reply hw/ip/aon_timer/data/aon_timer.hjson:221]: procedurally, this is what has happened: I added this tag in this PR, but @weicaiyang created a separate PR to fix this, so I rebased to pick his up, got a conflict, which I merged by choosing his fix. This is why the change shows up I guess in this PR, but it is a consequence of a merge. I can remove the merge, but this is sort of a git question in general.

File: hw/ip/csrng/rtl/csrng_cmd_stage.sv
Comment (mwbranstad) [reply hw/ip/csrng/rtl/csrng_cmd_stage.sv:325]: yeah, forgot to add this, it is done now.

File: hw/ip/csrng/data/csrng.hjson
Comment (mwbranstad) [reply hw/ip/csrng/data/csrng.hjson:225]: good catch...fix this wording as well as some related text.

**Review by vogelpi — APPROVED** [review]
LGTM!

File: hw/ip/aon_timer/data/aon_timer.hjson
Comment (vogelpi) [reply hw/ip/aon_timer/data/aon_timer.hjson]: Yes, the idea is not to use merge commits but just clean rebases. The commit of yours implementing that change then becomes empty and can be dropped. This has now happend, all good.

**Comment by mwbranstad** [comment]
@vogelpi Looks like this passes CI now, so we should be able to conclude on this one.

**Comment by vogelpi** [comment]
Thanks @mwbranstad for pinging me, I am merging this now.

**Pull request merged**
